flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/AutoRequirementDecrementingSlotPoolWrapper.java

```diff
@@ -0,0 +1,114 @@
+
+import org.apache.flink.runtime.clusterframework.types.AllocationID;
+import org.apache.flink.runtime.clusterframework.types.ResourceID;
+import org.apache.flink.runtime.clusterframework.types.ResourceProfile;
+import org.apache.flink.runtime.jobmanager.slots.TaskManagerGateway;
+import org.apache.flink.runtime.jobmaster.SlotInfo;
+import org.apache.flink.runtime.slots.ResourceRequirement;
+import org.apache.flink.runtime.taskexecutor.slot.SlotOffer;
+import org.apache.flink.runtime.taskmanager.TaskManagerLocation;
+
+import javax.annotation.Nullable;
+
+import java.util.Collection;
+import java.util.Optional;
+
+/**
+ * A wrapper around a {@link DeclarativeSlotPool} that automatically decrements requirements if a slot is freed or
+ * released.
+ */
+class AutoRequirementDecrementingSlotPoolWrapper implements DeclarativeSlotPool {
+
+	private final DeclarativeSlotPool backingSlotPool;
+
+	AutoRequirementDecrementingSlotPoolWrapper(DeclarativeSlotPool backingSlotPool) {
+		this.backingSlotPool = backingSlotPool;
+	}
+
+	@Override
+	public void increaseResourceRequirementsBy(ResourceCounter increment) {
+		backingSlotPool.increaseResourceRequirementsBy(increment);
+	}
+
+	@Override
+	public void decreaseResourceRequirementsBy(ResourceCounter decrement) {
+		backingSlotPool.decreaseResourceRequirementsBy(decrement);
+	}
+
+	@Override
+	public Collection<ResourceRequirement> getResourceRequirements() {
+		return backingSlotPool.getResourceRequirements();
+	}
+
+	@Override
+	public Collection<SlotOffer> offerSlots(Collection<? extends SlotOffer> offers, TaskManagerLocation taskManagerLocation, TaskManagerGateway taskManagerGateway, long currentTime) {
+		return backingSlotPool.offerSlots(offers, taskManagerLocation, taskManagerGateway, currentTime);
+	}
+
+	@Override
+	public Collection<SlotInfoWithUtilization> getFreeSlotsInformation() {
+		return backingSlotPool.getFreeSlotsInformation();
+	}
+
+	@Override
+	public Collection<? extends SlotInfo> getAllSlotsInformation() {
+		return backingSlotPool.getAllSlotsInformation();
+	}
+
+	@Override
+	public PhysicalSlot reserveFreeSlot(AllocationID allocationId, ResourceProfile requiredSlotProfile) {
+		return backingSlotPool.reserveFreeSlot(allocationId, requiredSlotProfile);
+	}
+
+	@Override
+	public Optional<ResourceCounter> freeReservedSlot(AllocationID allocationId, @Nullable Throwable cause, long currentTime) {
+		Optional<ResourceCounter> previouslyFulfilledRequirement = backingSlotPool.freeReservedSlot(allocationId, cause, currentTime);
+		previouslyFulfilledRequirement
+			.ifPresent(backingSlotPool::decreaseResourceRequirementsBy);
+		return previouslyFulfilledRequirement;
+	}
+
+	@Override
+	public ResourceCounter releaseSlots(ResourceID owner, Exception cause) {
+		ResourceCounter previouslyFulfilledRequirement = backingSlotPool.releaseSlots(owner, cause);
+
+		backingSlotPool.decreaseResourceRequirementsBy(previouslyFulfilledRequirement);
+		return previouslyFulfilledRequirement;
+	}
```

Comment:
Why are we decreasing the requirements if `releaseSlots` is called? I think this should be called if a TM's heartbeat times out.

Reply:
because we will request new slots for those that we just lost, and since this increments the requirements we have to decrement them somewhere to ensure the requirements stay the same.

Reply:
But shouldn't this be the responsibility of the user of this wrapper to do sort the requirements out correctly?

Reply:
Currently the user of this wrapper does not know what mapping the slot pool came up with.
We would have to include the originally matched profile in the slot information, that is returned for free slots.

Reply:
or return it in `reserveFreeSlot`

Reply:
Ah I see. This is definitely a recurring theme that the matching of the slots to required resources and the slots to `Executions` need to be aligned.

In the case of the `releaseSlots` method, I guess that the return value will tell us which requirements are no longer fulfilled.

Maybe it is not such a bad idea to return in some form the matched resource profile of a slot. This would allow the user to adjust the resource requirements if it decides to use this slot for a different `Execution`.

Reply:
hmm...it's not difficult to move the requirement adjustments into the wrapper, but the lack of updates to the slotpool-internal mappings could be problematic.

so let's see; we require one large and one small slot, get one each and do a perfect match.

```
requirements: S:1 L:1
fulfilledRequirements: S:1 L:1
slotToRequirementProfileMappings: S->S L->L
```

Now the scheduler makes things difficult, and uses L for S. We just adjust the requirements from the outside, and get:

```
requirements: S:2 L:0
fulfilledRequirements: S:1 L:1
mapping: S->S L->L
```

soo...we would now correctly determine that we need one more small slot, because we subtract the fulfilledRequirements from requirements, so that's fine.
But, if we release the large slot, then the wrapper needs to know what profile the slot was used for, to properly reduce the requirements. Currently this is based on what the slot pool returns, which would still based on the original mapping and as such would be incorrect.
I suppose if the scheduler were to use it then it would know for what requirement it was used for, so maybe the additional mapping would exist in the wrapper temporarily?...